profile2/tool/src/java/org/sakaiproject/profile2/tool/entityprovider/ProfileImageEntityProvider.java

```diff
@@ -0,0 +1,146 @@
+        return "profile-image";
+    }
+
+    @EntityCustomAction(action = "upload", viewKey = EntityView.VIEW_NEW)
+    public String upload(EntityView view, Map<String, Object> params) {
+
+        JSONObject result = new JSONObject();
+
+        result.put("status", "ERROR");
+
+        if (!checkCSRFToken(params)) {
+            return result.toJSONString();
+        }
+
+        User currentUser = UserDirectoryService.getCurrentUser();
+        String currentUserId = currentUser.getId();
+
+        if (currentUserId == null) {
+            LOG.warn("Access denied");
+            return result.toJSONString();
+        }
+
+        String mimeType = "image/png";
+        String fileName = UUID.randomUUID().toString();
+        String base64 = (String) params.get("base64");
+        byte[] imageBytes = Base64.decodeBase64(base64.getBytes());
+
+        if (imageLogic.setUploadedProfileImage(currentUserId, imageBytes, mimeType, fileName)) {
+            profileImageService.resetCachedProfileImageId(currentUserId);
+            result.put("status", "SUCCESS");
+        }
+
+        return result.toJSONString();
+    }
+
+    @EntityCustomAction(action = "details", viewKey = EntityView.VIEW_LIST)
+    public Object getProfileImage(OutputStream out, EntityView view, Map<String,Object> params) {
+
+        JSONObject result = new JSONObject();
+
+        result.put("status", "ERROR");
+
+        User currentUser = UserDirectoryService.getCurrentUser();
+        String currentUserId = currentUser.getId();
+
+        if (currentUserId == null) {
+            LOG.warn("Access denied");
+            return result.toJSONString();
+        }
+
+        String imageUrl = imageLogic.getProfileImageEntityUrl(currentUserId, ProfileConstants.PROFILE_IMAGE_MAIN);
+
+        result.put("url", imageUrl);
+        result.put("isDefault", imageLogic.profileImageIsDefault(currentUserId));
+        result.put("csrf_token", sessionManager.getCurrentSession().getAttribute("sakai.csrf.token"));
+        result.put("status", "SUCCESS");
+
+        return result.toJSONString();
+    }
+
+    @EntityCustomAction(action = "remove", viewKey = EntityView.VIEW_NEW)
+    public String remove(EntityView view, Map<String, Object> params) {
+
+        JSONObject result = new JSONObject();
+
+        result.put("status", "ERROR");
+
+        if (!checkCSRFToken(params)) {
+            return result.toJSONString();
+        }
+
+        User currentUser = UserDirectoryService.getCurrentUser();
+        String currentUserId = currentUser.getId();
+
+        if (currentUserId == null) {
+            LOG.warn("Access denied");
+            return result.toJSONString();
+        }
+
+        if (imageLogic.resetProfileImage(currentUserId)) {
+            profileImageService.resetCachedProfileImageId(currentUserId);
+            result.put("status", "SUCCESS");
+        }
+
+        return result.toJSONString();
+    }
+
+    private boolean checkCSRFToken(Map<String, Object> params) {
```

Comment:
@ern Agreed, but I haven't got the time to do that at the moment. One for the future, maybe when we reassess entity broker generally.